virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,284 @@ func (svc *googleService) insertSSHKey() error {
+}
+
+// detachDisk detaches a disk from an instance
+func (svc *googleService) detachDisk(disk *Disk) error {
+
+	var deviceName string
+
+	gDisk, err := svc.getDisk(disk.Name)
+	if err != nil {
+		return err
+	}
+	instance, err := svc.getInstance()
+	if err != nil {
+		return err
+	}
+
+	for _, attachedDisk := range instance.Disks {
+		if attachedDisk.Source == gDisk.SelfLink {
+			deviceName = attachedDisk.DeviceName
+			break
+		}
+	}
+
+	op, err := svc.service.Instances.DetachDisk(svc.vm.Project, svc.vm.Zone,
+		svc.vm.Name, deviceName).Do()
+
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+
+}
+
+// getFirewall gets details of a firewall
+func (svc *googleService) getFirewall() (*googlecloud.Firewall, error) {
+	return svc.service.Firewalls.Get(svc.vm.Project, svc.vm.Firewall).Do()
+}
+
+// addFirewallPorts adds new ports to a firewall
+func (svc *googleService) addFirewallPorts() error {
+
+	currFirewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	newRules := currFirewall.Allowed
+	for _, endpoint := range svc.vm.Endpoints {
+		newRules = append(newRules, &googlecloud.FirewallAllowed{
+			IPProtocol: endpoint.Protocol,
+			Ports:      endpoint.Ports,
+		})
+	}
+
+	return svc.patchFirewall(newRules)
+}
+
+// patchFirewall patches the firewall with the given allowed ports (rules)
+func (svc *googleService) patchFirewall(allowed []*googlecloud.FirewallAllowed) error {
+	newFirewall := &googlecloud.Firewall{
+		Allowed: allowed,
+	}
+	op, err := svc.service.Firewalls.Patch(svc.vm.Project, svc.vm.Firewall,
+		newFirewall).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForGlobalOperationReady(op.Name)
+}
+
+// removeFirewallPorts removes ports from a given firewall
+func (svc *googleService) removeFirewallPorts() error {
+	firewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	// Initialize new rules with the current one
+	// Refer to this link for structure of firewall resource
+	// https://cloud.google.com/compute/docs/reference/latest/firewalls
+	newRules := firewall.Allowed
+	// Define a map to identify ports to be removed
+	remPorts := make(map[string]map[string]bool)
+	// Define a map to deal with elements where the same protocol
+	// appears again
+	encountered := make(map[string]bool)
+	// Identify ports for each protocol to be removed
+	for _, endpoint := range svc.vm.Endpoints {
+		// Initialize the map only if the given protocol is encountered
+		// for the first time
+		if !encountered[endpoint.Protocol] {
+			remPorts[endpoint.Protocol] = make(map[string]bool)
+			encountered[endpoint.Protocol] = true
+		}
+		for _, port := range endpoint.Ports {
+			remPorts[endpoint.Protocol][port] = true
+		}
+	}
+
+	// Remove the the ports and protocols that are to be removed from the
+	// rules. Iterate through each item in the list of rules.
+	for indEndp := 0; indEndp < len(newRules); indEndp++ {
+		allowed := newRules[indEndp]
```

Comment:
"allowed" should be used going forward rather than dereferncing newRules[indEndp] again.

Reply:
Actually, we are modifying newRules as we traverse through the existing rules by removing the rules that are not needed. We are using allowed only at two places where we are okay with static value initialized at the beginning of the loop. Everywhere else, we need to refer to the modified newRules. Thus we cannot use allowed everywhere.